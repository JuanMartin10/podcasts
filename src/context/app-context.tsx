import { createContext, useContext, useState } from 'react';
import { useEntries } from '../hooks/useEntries';
import { type AppEntry } from '../models/types';

export interface AppContextTypes {
  loading: boolean;
  entries: AppEntry[] | undefined;
  currentPodcast: AppEntry | undefined;
  getEntries: () => Promise<void>;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setEntries: React.Dispatch<React.SetStateAction<AppEntry[] | undefined>>;
  setCurrentPodcast: React.Dispatch<React.SetStateAction<AppEntry | undefined>>;
}

interface ContextProps {
  children: React.ReactNode;
}

const AppContext = createContext<AppContextTypes | undefined>(undefined);

export const AppProvider: React.FC<ContextProps> = ({ children }) => {
  const [loading, setLoading] = useState(false);
  const [currentPodcast, setCurrentPodcast] = useState<AppEntry>();

  const { entries, getEntries, setEntries } = useEntries({
    loading,
    setLoading,
  });

  return (
    <AppContext.Provider
      value={{
        loading,
        entries,
        currentPodcast,
        getEntries,
        setLoading,
        setEntries,
        setCurrentPodcast,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextTypes => {
  const context = useContext(AppContext);
  if (context == null) {
    throw new Error('useAppContext must be used within an AppWrapper');
  }
  return context;
};
