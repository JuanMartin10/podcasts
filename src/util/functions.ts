import { Feed } from '../models/api-types';

export const formatTime = (ms: number) => {
  let totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  totalSeconds %= 3600;
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const secondsString = seconds.toString().padStart(2, '0');

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secondsString}`;
  } else {
    return `${minutes}:${secondsString}`;
  }
};

export const formatDate = (dateString: string) => {
  const date = new Date(dateString);

  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');

  return `${day}/${month}/${year}`;
};

export const formatApiEntries = (apiFeed: Feed) =>
  apiFeed.entry.map(entry => ({
    id: entry.id.attributes['im:id'],
    title: entry['im:name'].label,
    artist: entry['im:artist'].label,
    image: entry['im:image'][0].label,
    summary: entry.summary.label,
  }));
