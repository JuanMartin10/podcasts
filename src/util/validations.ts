export const isEmpty = (val: string | null | undefined) => {
  return val === '' || val === null || val === undefined;
};

export const isFalse = (val: boolean) => !val;
