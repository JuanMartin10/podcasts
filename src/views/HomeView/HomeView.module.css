.main {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: flex-start;
  flex-wrap: wrap;
  gap: 24px;
  margin-top: 24px;
}

.header {
  display: flex;
  justify-content: flex-end;
  align-items: flex-end;
}

.count {
  background-color: #508fcb;
  padding: 2px 4px;
  color: white;
  border-radius: 4px;
  font-weight: 700;
}
