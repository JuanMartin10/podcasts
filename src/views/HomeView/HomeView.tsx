import { useMemo } from 'react';
import SearchBar from '../../components/Searchbar/SearchBar';
import { useSearch } from '../../hooks/useSearch';
import Card from '../../components/Card/Card';
import { useAppContext } from '../../context/app-context';
import styles from './HomeView.module.css';

const HomeView = () => {
  const { entries, loading } = useAppContext();
  const { search: filterEntries, updateSearch: setFilterEntries } = useSearch();

  const filteredEntries = useMemo(() => {
    if (entries !== undefined) {
      return filterEntries !== undefined && entries.length > 0
        ? entries?.filter(el =>
            el.title.toLowerCase().includes(filterEntries.toLocaleLowerCase())
          )
        : entries;
    }
  }, [entries, filterEntries]);
  // debugger;
  console.log(typeof filteredEntries, '-');

  return (
    <div>
      <div className={styles.header}>
        {filteredEntries !== undefined && (
          <span className={styles.count}>{filteredEntries?.length}</span>
        )}
        <SearchBar
          value={filterEntries}
          onChangeEntries={(e: any) => {
            setFilterEntries(e.target.value);
          }}
        />
      </div>
      <div className={styles.main}>
        {filteredEntries === undefined ||
        (filteredEntries.length > 0 && loading) ? (
          <p>Cargando...</p>
        ) : (
          //  <p>pintando</p>
          filteredEntries?.map(en => <Card key={en.id} entry={en} />)
        )}
      </div>
    </div>
  );
};

export default HomeView;
