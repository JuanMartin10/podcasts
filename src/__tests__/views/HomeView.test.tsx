import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import {
  render,
  screen,
  cleanup,
  waitFor,
  fireEvent,
} from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import HomeView from 'views/HomeView/HomeView';
import { AppContext } from 'context/app-context';
import episodesJson from 'mocks/episodes.json';
import { formatApiEntries } from 'util/functions';

const renderWithProvider = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: false,
        currentPodcast: undefined,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );
const renderLoading = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: true,
        currentPodcast: undefined,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );

const wrapper = (
  <MemoryRouter>
    <Routes>
      <Route path='/' Component={HomeView} />
    </Routes>
  </MemoryRouter>
);

describe('<HomeView />', () => {
  beforeEach(() => {
    renderWithProvider(wrapper);
  });
  afterEach(cleanup);

  it('Should render', () => {
    renderWithProvider(wrapper);
  });

  it('Should match the snapshot', () => {
    const { container } = renderWithProvider(wrapper);
    expect(container).toMatchSnapshot();
  });

  it('Should render a header component', async () => {
    waitFor(() => {
      expect(screen.getByRole('header')).toHaveLength(1);
      expect(screen.getByText('Podcaster')).toBeDefined();
    });
  });
  it("Should render 'Loading...'", async () => {
    renderLoading(wrapper);
    await waitFor(() => {
      expect(screen.getByText(/loading/i)).toBeDefined();
    });
  });

  it('Should render images', async () => {
    await waitFor(() => {
      const images = screen.getAllByRole('img');
      expect(images).toBeDefined();
    });
  });

  it('Should filter searchbox', async () => {
    await waitFor(() => {
      const searchbox = screen.getByPlaceholderText(/filter podcasts/i);
      fireEvent.change(searchbox, { target: { value: 'joe' } });
      screen.getAllByText(/joe/i);
      expect(screen.queryByText(/elvis/i)).toBeNull();
    });
  });
});
