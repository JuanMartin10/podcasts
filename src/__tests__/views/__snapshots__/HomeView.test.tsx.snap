// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<HomeView /> > Should match the snapshot 1`] = `
<div>
  <div>
    <div
      class="_header_970eaf"
    >
      <span
        class="_count_970eaf"
      >
        4
      </span>
      <div
        class="_search_54eea8"
      >
        <input
          class="_searchTerm_54eea8"
          placeholder="Filter podcasts"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="_main_970eaf"
    >
      <a
        href="/podcast/1535809341"
        style="text-decoration: none;"
      >
        <div
          class="_item_1f2933"
        >
          <div
            class="_image_1f2933"
          >
            <img
              alt="The Joe Budden Podcast-https://is3-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/55x55bb.png"
              src="https://is3-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/55x55bb.png"
            />
          </div>
          <div
            class="_info_1f2933"
          >
            <div
              class="_titles_1f2933"
            >
              <h3
                class="_title_1f2933"
              >
                The Joe Budden Podcast
              </h3>
              <p
                class="_author_1f2933"
              >
                Author: The Joe Budden Network
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/podcast/1678378239"
        style="text-decoration: none;"
      >
        <div
          class="_item_1f2933"
        >
          <div
            class="_image_1f2933"
          >
            <img
              alt="This Little Light-https://is3-ssl.mzstatic.com/image/thumb/Podcasts116/v4/f6/cf/66/f6cf66f5-8696-1903-f3b3-e95c507c8e9c/mza_5620205623101925377.jpg/55x55bb.png"
              src="https://is3-ssl.mzstatic.com/image/thumb/Podcasts116/v4/f6/cf/66/f6cf66f5-8696-1903-f3b3-e95c507c8e9c/mza_5620205623101925377.jpg/55x55bb.png"
            />
          </div>
          <div
            class="_info_1f2933"
          >
            <div
              class="_titles_1f2933"
            >
              <h3
                class="_title_1f2933"
              >
                This Little Light
              </h3>
              <p
                class="_author_1f2933"
              >
                Author: Cadence13 and Parallel
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/podcast/1096830182"
        style="text-decoration: none;"
      >
        <div
          class="_item_1f2933"
        >
          <div
            class="_image_1f2933"
          >
            <img
              alt="Drink Champs-https://is4-ssl.mzstatic.com/image/thumb/Podcasts113/v4/46/39/69/46396977-6d87-f120-6432-6373dcefdf9b/mza_6589930413574180525.jpg/55x55bb.png"
              src="https://is4-ssl.mzstatic.com/image/thumb/Podcasts113/v4/46/39/69/46396977-6d87-f120-6432-6373dcefdf9b/mza_6589930413574180525.jpg/55x55bb.png"
            />
          </div>
          <div
            class="_info_1f2933"
          >
            <div
              class="_titles_1f2933"
            >
              <h3
                class="_title_1f2933"
              >
                Drink Champs
              </h3>
              <p
                class="_author_1f2933"
              >
                Author: Interval Presents
              </p>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/podcast/1460157002"
        style="text-decoration: none;"
      >
        <div
          class="_item_1f2933"
        >
          <div
            class="_image_1f2933"
          >
            <img
              alt="Million Dollaz Worth Of Game-https://is1-ssl.mzstatic.com/image/thumb/Podcasts125/v4/7b/cf/f6/7bcff6bb-5f99-6c2f-c6c5-3a9799f3df21/mza_8544742664200824246.jpg/55x55bb.png"
              src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts125/v4/7b/cf/f6/7bcff6bb-5f99-6c2f-c6c5-3a9799f3df21/mza_8544742664200824246.jpg/55x55bb.png"
            />
          </div>
          <div
            class="_info_1f2933"
          >
            <div
              class="_titles_1f2933"
            >
              <h3
                class="_title_1f2933"
              >
                Million Dollaz Worth Of Game
              </h3>
              <p
                class="_author_1f2933"
              >
                Author: Barstool Sports
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
